### PR TITLE
Add RSA support to DNSSEC infra via OpenSSL

### PR DESCRIPTION
**Pull request by rgacogne**

### CODE REVIEW

**Comment by rgacogne** [comment]
Fixed space nits reported by @zeha (thx!)

**Comment by zeha** [comment]
LGTM, maybe someone else could review this, too? (@mind04?)

File: pdns/opensslsigners.cc [hunk 1]
```diff
@@ -5,9 +5,306 @@
...
 #include "dnssecinfra.hh"
 
+class OpenSSLRSADNSCryptoKeyEngine : public DNSCryptoKeyEngine
+{
+public:
+  explicit OpenSSLRSADNSCryptoKeyEngine(unsigned int algo) : DNSCryptoKeyEngine(algo)
+  {
+    int ret = RAND_status();
+    if (ret != 1) {
+      throw runtime_error(getName()+" insufficient entropy");
```
Comment (cmouse) [inline pdns/opensslsigners.cc]: We just die if we are low on entropy?

File: pdns/opensslsigners.cc [hunk 2]
```diff
@@ -5,9 +5,306 @@
...
+
+  if (hash.size() == SHA_DIGEST_LENGTH) {
+    hashKind = NID_sha1;
+  }
+  else if (hash.size() == SHA256_DIGEST_LENGTH) {
+    hashKind = NID_sha256;
+  }
+  else {
+    hashKind = NID_sha512;
+  }
```
Comment (cmouse) [inline pdns/opensslsigners.cc]: no sha384?

File: pdns/opensslsigners.cc [hunk 3]
```diff
@@ -5,9 +5,306 @@
...
+  unsigned char signature[RSA_size(d_key)];
+  unsigned int signatureLen = 0;
+
+  if (hash.size() == SHA_DIGEST_LENGTH) {
+    hashKind = NID_sha1;
+  }
+  else if (hash.size() == SHA256_DIGEST_LENGTH) {
+    hashKind = NID_sha256;
+  }
+  else {
```
Comment (cmouse) [inline pdns/opensslsigners.cc]: should we ensure it is valid size here?

Comment (mind04) [inline pdns/opensslsigners.cc]: why, there are only 3 hash types (possible sizes) in this->hash

File: pdns/opensslsigners.cc [hunk 4]
```diff
@@ -5,9 +5,306 @@
...
+  const unsigned char* raw = (const unsigned char*)input.c_str();
+
+  if (raw[0] != 0) {
+    exponent = input.substr(1, raw[0]);
+    modulus = input.substr(raw[0]+1);
+  } else {
+    exponent = input.substr(3, raw[1]*0xff + raw[2]);
+    modulus = input.substr(3+ raw[1]*0xff + raw[2]);
+  }
+
```
Comment (mind04) [inline pdns/opensslsigners.cc]: possibility for reading out of bounds and an uncaught out_of_range exception from substr() if input  is not big enough

File: pdns/opensslsigners.cc [hunk 5]
```diff
@@ -5,9 +5,306 @@
...
+  places["Exponent1"]=&d_key->dmp1;
+  places["Exponent2"]=&d_key->dmq1;
+  places["Coefficient"]=&d_key->iqmp;
+
+  drc.d_algorithm = pdns_stou(stormap["algorithm"]);
+
+  string raw;
+  for(const places_t::value_type& val :  places) {
+    raw=stormap[toLower(val.first)];
+    *val.second = BN_bin2bn((unsigned char*) raw.c_str(), raw.length(), NULL);
```
Comment (mind04) [inline pdns/opensslsigners.cc]: it is possible to reuse the existing  *val.second structure (s/NULL/*val.second)
Maybe we even leak memory here.

File: pdns/opensslsigners.cc [hunk 6]
```diff
@@ -5,9 +5,306 @@
...
+
+  if (raw[0] != 0) {
+    exponent = input.substr(1, raw[0]);
+    modulus = input.substr(raw[0]+1);
+  } else {
+    exponent = input.substr(3, raw[1]*0xff + raw[2]);
+    modulus = input.substr(3+ raw[1]*0xff + raw[2]);
+  }
+
+  d_key->e = BN_bin2bn((unsigned char*)exponent.c_str(), exponent.length(), NULL);
```
Comment (mind04) [inline pdns/opensslsigners.cc]: reuse BIGNUM

File: pdns/opensslsigners.cc [hunk 7]
```diff
@@ -5,9 +5,306 @@
...
+  if (!d_key->e) {
+    throw runtime_error(getName()+" error loading e value of public key");
+  }
+  d_key->n = BN_bin2bn((unsigned char*)modulus.c_str(), modulus.length(), NULL);
```
Comment (mind04) [inline pdns/opensslsigners.cc]: reuse BIGNUM

**Comment by mind04** [comment]
and maybe clear the private parts of the key structure in fromPublicKeyString()

**Comment by rgacogne** [comment]
Fixed most comments. I don't really know what else to do when we lack entropy. As for the private parts clearing, does it make sense to call fromPublicKeyString() on a signer having created or loaded a private key already?

**Comment by zeha** [comment]
I doubt one can run out of openssl entropy except if everything went wrong, and the ECDSA implementation also throws here.

**Comment by mind04** [comment]
There is no guarantee that a signer is used for a single task with a single key. Removing key parts that belong to a possibly different key pair on update seems like the right thing to do, to me. 

The entropy throw already saved us once. Please don't touch it ;)

**Comment by rgacogne** [comment]
Ok, we now completely reset the key on from* calls

**Pull request merged**